server/vulnerabilities/cpe.go

```diff
@@ -139,67 +144,312 @@ func cleanAppName(appName string) string {
 	return strings.TrimSuffix(appName, ".app")
 }
 
-var onlyAlphaNumeric = regexp.MustCompile("[^a-zA-Z0-9]+")
-
-func CPEFromSoftware(db *sqlx.DB, software *fleet.Software) (string, error) {
-	targetSW := ""
-	switch software.Source {
-	case "apps":
-		targetSW = "macos"
-	case "python_packages":
-		targetSW = "python"
-	case "chrome_extensions":
-		targetSW = "chrome"
-	case "firefox_addons":
-		targetSW = "firefox"
-	case "safari_extensions":
-		targetSW = "safari"
-	case "deb_packages":
-	case "portage_packages":
-	case "rpm_packages":
-	case "npm_packages":
-		targetSW = `"node.js"`
-	case "atom_packages":
-	case "programs":
-		targetSW = `"windows*"`
-	case "ie_extensions":
-	case "chocolatey_packages":
+var onlyAlphaNumeric = regexp.MustCompile(`[^a-zA-Z0-9]+`)
+
+// sanitizeMatch sanitizes the search string for sqlite fts queries. Replaces all special characters with spaces.
+func santizeMatch(s string) string {
+	return onlyAlphaNumeric.ReplaceAllString(s, " ")
+}
+
+var sanitizeVersionRe = regexp.MustCompile(`[^a-zA-Z0-9_-]+`)
+
+// sanitizeVersion attempts to sanitize versions and attempt to make it dot separated.
+// Eg Zoom reports version as "5.11.1 (8356)". In the NVD CPE dictionary it should be 5.11.1.8356.
+func sanitizeVersion(version string) string {
+	parts := onlyAlphaNumeric.Split(version, -1)
+	return strings.Trim(strings.Join(parts, "."), ".")
+}
+
+// TODO: add more vendors
+var macOSVendors = map[string]string{
+	"com.postmanlabs.mac":           "getpostman",
+	"org.virtualbox.app.VirtualBox": "oracle",
+}
+
+const cpeTranslationsFilename = "cpe_translations.json"
+
+func loadCPETranslations(path string) (CPETranslations, error) {
+	f, err := os.Open(path)
+	if err != nil {
+		return nil, err
+	}
+	defer f.Close()
+
+	var translations CPETranslations
+	if err := json.NewDecoder(f).Decode(&translations); err != nil {
+		return nil, fmt.Errorf("decode json: %w", err)
+	}
+
+	return translations, nil
+}
+
+func DownloadCPETranslations(vulnPath string, client *http.Client) error {
+	ghClient := github.NewClient(client)
+
+	opts := &github.RepositoryContentGetOptions{
+		Ref: "michal-6628-macos-vuln",
+	}
+	r, resp, err := ghClient.Repositories.DownloadContents(context.Background(), "fleetdm", "nvd", cpeTranslationsFilename, opts)
```

Comment:
Could you do `fContent, _, resp, err := ghClient.Repositories.GetContents(context.Background(), "fleetdm", "nvd", cpeTranslationsFilename, opts)` here instead? Then you could something like `download.Download(fContent.GetDownloadURL(),...)` that should simplify this method by a lot.

Reply:
`GetContents` retrieves the file metadata AND its contents, meaning it has already downloaded the file. Using `download.Download` afterwards would mean downloading the file twice. While it would be simpler to reuse the same code, I don't think we should download twice.

Reply:
Ah right, to download the metadata only you will need to target the containing directory, not the file, so maybe something like:
```
_, dirEntries, resp, err := ghClient.Repositories.GetContents(context.Background(), "fleetdm", "nvd", ".", opts)
// Find entry for cpeTranslationsFilename
download.Download(.., matchingEntry.GetDownloadURL(),...)
```